### ISSUE TITLE
Updating transitive UnitsNet dependency cause wrong unit

### ISSUE THREAD

**Issue by MarS0K** [angularsen/UnitsNet#1068]
**Describe the bug**
I have nuget project A that uses UnitsNet 4.107.0. It uses Length quantity and LengthUnit.Meter to do some work (serialize Length property to double in meters).
Then I have application project B that uses UnitsNet 4.126.0. It uses classes from project A. Instead of Meters it produce MegalightYear.

**To Reproduce**
Project A code (4.107.0):
```
public static class QuantityProcessor
    {
        private const double TenMeters = 10;
        private const LengthUnit MeterUnit = LengthUnit.Meter;

        public static double GetTenMetersFromMeterUnit()
            => Length.From(TenMeters, MeterUnit).Meters;

        public static double GetTenMetersAsMeterUnit()
            => Length.FromMeters(TenMeters).As(MeterUnit);
    }
```
Project B code (4.126.0):
```
public static class AppQuantityProcessor
    {
        public static double GetTenMetersFromMeterUnit()
            => QuantityProcessor.GetTenMetersFromMeterUnit();

        public static double GetTenMetersAsMeterUnit()
            => QuantityProcessor.GetTenMetersAsMeterUnit();
    }
```
Test code that uses project B:
```
public class AppQuantityProcessorTest
    {
        private const double TenMeters = 10;

        [Fact]
        public void GetTenMetersAsMeterUnit_ShouldReturnTenMeters()
        {
            Assert.Equal(TenMeters, AppQuantityProcessor.GetTenMetersAsMeterUnit());
        }

        [Fact]
        public void GetTenMetersFromMeterUnit_ShouldReturnTenMeters()
        {
            Assert.Equal(TenMeters, AppQuantityProcessor.GetTenMetersFromMeterUnit());
        }
    }
```
Methods that library B used return values in megalight years.

**Expected behavior**
Methods that library B used return values in meters.

New units should be appended to Enum, not inserted in alphabetical order. So LengthUnit.Meter should always be same value:
UnitsNet 4.107.0 - 16
UnitsNet 4.126.0 - 18

Every new unit can introduce breaking change.
Is it possible to append to end all new units instead? I don't want to update UnitsNet dependency in every library that not need new units.

**Comment by tmilnthorp** [comment]
I agree we should append to the end of the list, and could enforce a rule that new units go at the end of the json files. 

Where this will get hard is the prefixes. They’re auto generated and additions or removals would still change the order. The only way I could think around that is to add some sort of “referencing prefix” where you’d add a prefix to the end of the file that references an existing entry higher in the file. 

Removals would have to be obsoleted for the next major rev.

**Comment by remcoros** [comment]
This just hit us very hard as well.

We have mappers of one unit type of library X (commercial library) to UnitsNet unit types.

Due to a transitive update of the UnitsNet dependency, this broke our code where for example "Millimeter" was mapped to "Mil" and "Meter" to "Lightyear".

This is really bad. Enum value order should definitely NOT change between minor versions of UnitsNet in my opinion. And in case this changes between major versions, it warrants a big warning.

**Comment by angularsen** [comment]
Thanks for reporting. It makes sense now that you report it, transitive dependencies can be a pain.

I don't immediately see a good approach to ensuring enum values are only appended due to the code-generator we are using. We would have to somehow make the codegen aware of the latest generated values.
If you have any good ideas I'm all ears. I'll have to think about it.

**Comment by angularsen** [comment]
Some possible approaches that come to mind:
- Codegen scans the existing generated unit enum source files and maps out what already exists, before generating new code
- Codegen compiles UnitsNet and looks up the existing types before generating new code (slow)

**Comment by Muximize** [comment]
How about adding an order or position number to each unit in the definition file, a bit like Protobuf Field Numbers:

```
  "Units": [
    {
      "SingularName": "Meter",
      "PluralName": "Meters",
      "EnumOrder": 1
      ...
    },
    {
      "SingularName": "Mile",
      "PluralName": "Miles",
      "EnumOrder": 2
      ...
```
Or is that too much of an implementation detail at this level? (Also, naming could be better?)

**Comment by angularsen** [comment]
Prefixes kilo, mega etc is still unsolved, plus I really want to avoid having to manually write out each index for every 1000+ units and future units.

**Comment by Muximize** [comment]
Ah yes, didn't think of that.

How about giving all possible prefixes a "global" index (in `JsonTypes.Prefix`?), then when generating the unit enums with their relevant prefixes use this global index as an explicit value.

Would be a breaking change because the indexes probably change for lots of units.

**Comment by angularsen** [comment]
@Muximize It would still only solve the prefix problem, and I really want to avoid manually specifying indexes in JSON as it makes it more difficult to add new units for collaborators when you have to find the next available number, plus parallel pull requests on the same quantity will get nasty merge conflicts with room for error we would have to guard against.

It just feels like the wrong solution.

I'm thinking the code generator can do a two-pass thing, where it first scans what enum indexes are already populated in some file, then it sequentially adds index values to any enums without one.

This way the enums can still be sorted alphabetically, but they are assigned unique and strictly increasing integer values.

**Comment by angularsen** [comment]
Hi all, I believe I found a fix for this problem.
Merged in #1097 .

Essentially it tries to allocate enum values and keeps track of those in a separate JSON file.
It will automatically update the JSON file as part of pull requests that add new units, when they run code generation.
This means it requires no extra steps for contributors adding units, which I think is important for this to work well.

I tried a couple of scenarios and it seems to work as intended.
- Adding one or more units to a quantity, allocates enum values that are unique to that quantity.
- The allocations pick randomly from the next 10 available values to simplify merge conflicts on parallel pull requests adding units to the same quantity
- Adding more than 10 units to a quantity works, it allocates random values in batches of 10.
- The allocations fills any gaps in the sequence, since we create gaps by randomly picking enum values.
- Removing a unit does not remove the allocation, so the value is reserved forever 

You can read more in the PR if you are interested and let me know if you have feedback on the solution.

Starting from now and out, the enum values should stay consistent 🤞
